platform/solr/solr-factory/src/test/groovy/org/codice/solr/factory/impl/SolrClientAdapterAsyncSpec.groovy

```diff
@@ -0,0 +1,1253 @@
+        run(_) >> Stub(FailsafeFuture)
+      }
+    }
+  }
+
+  @Unroll
+  def 'test constructor when client is created after #created_after and becomes available'() {
+    given: "a client and creator to verify that failsafe actually delegates to our creator and client"
+      def client = Mock(SolrClient) {
+        1 * ping() >> pingOkResponse
+      }
+      def creator = Mock(Creator) {
+        1 * create() >> client
+      }
+
+    and: "controllers that will actually proceed through Solr"
+      def createController = new FailsafeController('SolrClient Creation')
+      def pingController = new FailsafeController('SolrClient Ping').onNextExecution(doProceed())
+
+      if (return_or_throw == SolrClient) {
+        createController.onNextExecution(doProceed())
+      } else {
+        // null or exception
+        createController.onNextExecution {
+          doThrowOrReturn(return_or_throw)
+              .then().doProceed()
+        }
+      }
+
+    when: "creating an adapter"
+      def adapter = new SolrClientAdapter(CORE, creator, createController.&with, pingController.&with)
+
+    then: "the reported core should correspond to the requested one"
+      adapter.core == CORE
+
+    and:
+      "our creator should be registered with the adapter"
+      adapter.creator.is(creator)
+
+    when: "waiting for failsafe to complete"
+      createController.waitForSuccessfulCompletion()
+      pingController.waitForSuccessfulCompletion()
+
+    then: "the adapter should be available"
+      adapter.isAvailable()
+
+    and: "the underlying client should be the one we created"
+      adapter.proxiedClient.is(client)
+
+    and: "the underlying client should never be closed"
+      0 * client.close()
+
+    and: "there should be no registered initializers or listeners"
+      !adapter.hasInitializers() && !adapter.hasListeners()
+
+    and: "verify failsafe controllers"
+      createController.verify()
+      pingController.verify()
+
+    cleanup:
+      createController?.shutdown()
+      pingController?.shutdown()
+
+    where:
+      // return_or_throw -> null to return null, exception to be thrown, SolrClient to return a client
+      created_after                                                    || return_or_throw
+      '1 attempt'                                                      || SolrClient
+      '2 attempts where the first one returned null'                   || null
+      '2 attempts where the first one failed with SolrServerException' || SolrServerException
+      '2 attempts where the first one failed with SolrException'       || new SolrException(
+          ErrorCode.UNKNOWN, 'testing')
+      '2 attempts where the first one failed with IOException'         || IOException
+      '2 attempts where the first one failed with RutimeException'     || RuntimeException
+  }
+
+  @Unroll
+  // the following needs to be changed as it should be closing the whole thing and not re-execute the creation
```

Comment:
I was trying to get someones opinion. Although right now, it ain't an issue as this would only happen during shutdown of the thread pool and the thread pool is never shutdown anyway (as it wasn't in the original code where it was moved from). Until such time where we do implement proper shutdown, we can keep the code as it is. As for opening up another ticket, my issue is that it will only be closed later anyway.